hieradata_aws/class/integration/jenkins.yaml

```diff
@@ -58,6 +59,11 @@ govuk_jenkins::jobs::ask_export::folder_id_third_party: '1RH44n_9Ps_Syeq7o-xJFuF
 govuk_jenkins::jobs::ask_export::s3_bucket_name_gcs_public_questions: 'govuk-uk-integration-ask-export'
 govuk_jenkins::jobs::ask_export::run_daily: false
 
+govuk_jenkins::jobs::search_api_learn_to_rank::train_instance_type: ml.c5.xlarge
+govuk_jenkins::jobs::search_api_learn_to_rank::train_instance_count: 2
```

Comment:
Ah maybe this need clarifying. This `train_instance_count` variable refers to the number of instance used by SageMaker to train the model.

The LTR pipeline contains three distinct steps:
1. collect and preprocess training, validation datasets
2. training the model
3. deploying the model

Step 1 (preprocessing datasets) is done by SSHing into a [vanilla EC2 instance created in the LTR ASG](https://github.com/alphagov/search-api/blob/34b67fc3a150b64044c025e7eea4d195d213bf42/ltr/jenkins/start.sh#L18) and running a script. Then [step 2](https://github.com/alphagov/search-api/blob/34b67fc3a150b64044c025e7eea4d195d213bf42/ltr/jenkins/train.py#L29) and [step 3](https://github.com/alphagov/search-api/blob/34b67fc3a150b64044c025e7eea4d195d213bf42/ltr/jenkins/deploy.py#L88) happen by API calls to SageMaker, which itself spins up other instances to do those things (but that's all managed).

We do only spin one EC2 instance to collect and preprocess the data (i.e. the rake tasks). However to train the model we use this variable to tell SageMaker to use multiple instances (which I assume speeds up the process).

I'll update the descriptions for the variables. Let me know if that still doesn't make sense.

---
That aside - what's changed on Jenkins is where the API calls to SageMaker are made. Previously on concourse, the SageMaker API calls where made from the concourse runner container. Now with Jenkins, the are made from the EC2 instance created from step 1 after the preprocessing of the datasets (instead of directly Jenkins itself).

#### Previously with concourse
1. Concourse runner container --(SSH)-> LTR EC2 Instance -> Preprocess datasets
2. Concourse runner container -> SageMaker Train API call ~> SageMaker spins up some instances to train
3. Concourse runner container -> SageMaker Deploy API call ~> SageMaker spins up some instances to deploy

#### Now with Jenkins
1. Jenkins --(SSH)-> LTR EC2 -> Preprocess datasets
2. LTR EC2 (same SSH session) -> SageMaker Train API call ~> SageMaker spins up some instances to train
3. LTR EC2 (same SSH session) -> SageMaker Deploy API call ~> SageMaker spins up some instances to deploy

Reply:
Thanks @theseanything, this is super helpful - it's been ages since I've looked at this codebase.

> To train the model we use this variable to tell SageMaker to use multiple instances (which I assume speeds up the process).

I think this might be an improvement for the deploy job. We'll be serving HTTP calls with more instances, so perhaps that will translate into lower latencies - would be cool to monitor this and record the effect somewhere.

Using two instances won't improve the speed of the training. Instead this will result in training two different versions of the same model. The training script is a single threaded process which won't make use of multiple machines, sadly. Though, this would be a really cool thing to work on, if you're interested on making that happen - if we wanted to train on _all_ the data we have available rather than a subset this would be useful to consider.

Reply:
Ahh interesting, reading the AWS docs it wasn't clear what happens when instance_count > 1 for training. For sure, it'd be nice to see if could move to distributed train model. I probably won't get time myself to look at it.

Reply:
Cool, thanks for looking into it. Definitely excited to see whether increasing the api instance count will have an impact on performance. Let's change it to `1` here for the training instances and we should be good 👍